compose: Add stream privacy decorations in banner box.

srdeotarse opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR is for issue #19878 
Fixes part (1/2) -
Add stream privacy decorations in banner box.

Part (2/2) in another PR as suggested by @alya in Zulip chat https://chat.zulip.org/#narrow/stream/9-issues/topic/missing.20hash.20before.20stream.20name/near/1261461 -
Add stream privacy decorations in compose box.

**Testing plan:** <!-- How have you tested? -->
Lint test.
Node test.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![image](https://user-images.githubusercontent.com/53873549/145938338-0371369c-5c06-418a-ac88-b43c377f1df7.png)

**Changes explained:**

Earlier sent message notification - "Sent! Your message is outside current narrow. Narrow to test > blue"

Updated sent message notification - "Sent! Your message is outside current narrow. Narrow to #test > blue"


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

alya commented:
Hm, that hash looks overly large to me.

Could you also:
1) Add screenshots for a private stream and a world public stream?
2) Make sure your PR passes checks.

Thanks!

srdeotarse commented:
> Hm, that hash looks overly large to me.
> 
> Could you also:
> 
> 1. Add screenshots for a private stream and a world public stream?
> 2. Make sure your PR passes checks.
> 
> Thanks!

Yes, I will adjust the hash size and also add screenshots for mentioned cases.
I will make sure that my PR passes checks as well.

Thanks for reviewing my PR.

srdeotarse commented:
> Hm, that hash looks overly large to me.
> 
> Could you also:
> 
> 1. Add screenshots for a private stream and a world public stream?
> 2. Make sure your PR passes checks.
> 
> Thanks!

Fix vertical alignment and size of stream privacy decoration icons in banner box.

Screenshots-

Public stream -
![image](https://user-images.githubusercontent.com/53873549/146320496-6396d3d3-049b-446a-bb52-9272a38c1838.png)

Private stream -
![image](https://user-images.githubusercontent.com/53873549/146320563-6af95b18-2db6-4821-87e4-661cfb21d3c2.png)

Web Public stream-
![image](https://user-images.githubusercontent.com/53873549/146320664-8fb8f7e8-c217-4203-9221-c748274fa11e.png)

raghupalash commented:
Thanks for working on this @srdeotarse! I have a review regarding the commits.
Your commit messages don't comply with the [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-discipline) guide.

Also you might want to clean up your commit history, check out - https://zulip.readthedocs.io/en/latest/git/fixing-commits.html.

Goodday! 😸 🐋

srdeotarse commented:
@raghupalash I have cleaned my commit history and changed commit messages according to commit guidelines.

alya commented:
Thanks! It looks to me like there's not enough spacing to the left of the lock and globe icons -- there should be as much space as you normally have between words.

For the hash, can we try having less space between the # and the name of the stream?

srdeotarse commented:
> Thanks! It looks to me like there's not enough spacing to the left of the lock and globe icons -- there should be as much space as you normally have between words.
> 
> For the hash, can we try having less space between the # and the name of the stream?

Adjusted spacing to the left of globe and lock icons in banner box. Also, reduced spacing between # and name of stream.

alya commented:
@srdeotarse  Could you add updated screenshots here? Thanks!

srdeotarse commented:
@alya  Adding the updated screenshots -

Added spacing to the left of lock icon - 
![image](https://user-images.githubusercontent.com/53873549/149086049-bd4ae3f6-ff2c-48e9-a1fa-88bb464f864f.png)

Added spacing to the left of globe icon - 
![image](https://user-images.githubusercontent.com/53873549/149086216-468696b2-696f-4cd1-a4b7-ecb5627387f5.png)

Reduced spacing between hash and stream name -
![image](https://user-images.githubusercontent.com/53873549/149086374-a84f4da7-5d48-4a0d-b4aa-0503f7fa9a5a.png)

raghupalash commented:
> @raghupalash I have cleaned my commit history and changed commit messages according to commit guidelines.

I think you'd want to squash all the commits in one single commit (as they all belong to a single change). This might help https://zulip.readthedocs.io/en/latest/git/fixing-commits.html#fixing-the-last-commit. Feel free to bump a thread on the development-help stream if you can't figure it out.

raghupalash commented on static/js/stream_data.js:
```diff
@@ -571,6 +571,14 @@ export function get_stream_privacy_policy(stream_id) {
     return stream_privacy_policy_values.private_with_public_history.code;
 }
 
+export function get_invite_only(stream_name) {
```
`is_invite_only_by_stream_name` function on line 582 does the exact same thing as this, infact `is_invite_only_by_stream_name` was formerly named as `get_invite_only` and renamed by Tim on commit [3f4d66109be0265a45713419bbee14394c98414f](https://github.com/zulip/zulip/commit/3f4d66109be0265a45713419bbee14394c98414f). So I think we should use the already available function instead of making a new one.

srdeotarse replied:
I have removed `get_invite_only` function and used `is_invite_only_by_stream_name`.

raghupalash commented on static/js/notifications.js:
```diff
@@ -635,17 +643,18 @@ export function notify_local_mixes(messages, need_user_to_scroll) {
-            {defaultMessage: "Narrow to {message_recipient}"},
-            {message_recipient: get_message_header(message)},
-        );
+        const default_message = "Narrow to";
```
I think `$t` was used here for translation purposes, so I think we should still stick with using it, plus I don't think your implementation requires you to remove it anyways, you can just let that line be and use it in the `compose_notification` template as it was used before with your additional changes.

srdeotarse replied:
Now I have used $t along with my changes.

alya commented:
@amanagr perhaps you could review and/or finish up this PR as well?

amanagr commented on static/styles/compose.css:
```diff
@@ -150,6 +149,22 @@
     }
 }
 
+#compose-lock-icon,
+.hashtag {
+    vertical-align: middle;
+}
+
+#compose-lock-icon,
+#compose-globe-icon {
+    position: relative;
+    margin-left: 3px;
+    width: 0;
+}
+
+.hashtag {
+    margin-right: -2px;
+}
+
```
These need to be scoped inside `.compose-notifications-content`

amanagr committed:
compose: Convert globe and lock icon identifiers to class.

We will be using them in multiple places inside compose so it
makes sense to have them as a class instead of an id.

srdeotarse committed:
compose: Add decoration for public stream in stream name box.

srdeotarse committed:
compose: Add stream privacy decorations in banner box.

amanagr commented:
I removed the underline on link hover since it was not playing nicely with the icons. It just has an increased opacity effect on hover now.
<img width="855" alt="Screenshot 2022-06-29 at 3 16 52 PM" src="https://user-images.githubusercontent.com/25124304/176410160-1dd13aa8-234b-46a8-b632-18b408bd6e6d.png">

<img width="855" alt="Screenshot 2022-06-29 at 3 16 22 PM" src="https://user-images.githubusercontent.com/25124304/176410137-d9296625-990d-46fe-bd4c-292913d30a17.png">
<img width="855" alt="Screenshot 2022-06-29 at 3 16 37 PM" src="https://user-images.githubusercontent.com/25124304/176410168-4737438f-60f9-4c3e-812a-f26c749136b9.png">
<img width="855" alt="Screenshot 2022-06-29 at 3 16 29 PM" src="https://user-images.githubusercontent.com/25124304/176410179-f99ac9fb-e850-4b65-bb0c-24b3dc290da9.png">

srdeotarse commented:
@amanagr I have changed the author of the last two commits to me.

amanagr commented:
@srdeotarse did you review the changes I made? Do they look good to you?

alya commented:
Thanks for updating this!

I'm not sure about the globe and lock, but the font on `#` looks odd to me. Could we make it less bold to match the rest of the text?

alya commented:
Seeing the icons added, it now feels odd not to have the appropriate stream privacy icon in a couple of other places in the compose box area:

- The "Message..." button
![Screen Shot 2022-07-07 at 10 44 12 AM](https://user-images.githubusercontent.com/2090066/177837215-4b8b986f-ae7f-49f8-bd78-d35b924e066b.png)

- The "Message..." placeholder text
![Screen Shot 2022-07-07 at 10 44 19 AM](https://user-images.githubusercontent.com/2090066/177837080-7cf6da1b-32b5-4757-866c-62f4ecdc9658.png)

Would it be possible to make these be based on the stream privacy setting as well?

alya commented:
@srdeotarse , will you have time to address the remaining feedback on this PR, or should @amanagr take over from here?

alya commented:
Let's just complete the banner icon changes in this PR.
